rclpy/src/rclpy/_rclpy.c

```diff
@@ -317,9 +320,70 @@ _rclpy_parse_args(PyObject * pyargs, rcl_arguments_t * parsed_args)
   ret = rcl_parse_arguments(num_args, const_arg_values, allocator, parsed_args);
 
   if (ret != RCL_RET_OK) {
-    PyErr_Format(PyExc_RuntimeError, "Failed to init: %s", rcl_get_error_string().str);
+    if (ret == RCL_RET_INVALID_ROS_ARGS) {
+      PyErr_Format(
+        RCLInvalidROSArgsError,
+        "Failed to parse arguments: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      goto cleanup;
+    }
+    PyErr_Format(
+      PyExc_RuntimeError,
+      "Failed to init: %s",
+      rcl_get_error_string().str);
     rcl_reset_error();
+    goto cleanup;
+  }
+
+  int unparsed_ros_args_count = rcl_arguments_get_count_unparsed_ros(parsed_args);
+  if (unparsed_ros_args_count > 0) {
+    int * unparsed_ros_args_indices = NULL;
+    ret = rcl_arguments_get_unparsed_ros(
+      parsed_args, allocator, &unparsed_ros_args_indices);
+    if (ret != RCL_RET_OK) {
+      PyErr_Format(
+        PyExc_RuntimeError,
+        "Failed to get unparsed ROS arguments: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      goto cleanup;
+    }
+
+    PyObject * unknown_ros_pyargs = NULL;
+
+    pyargs = PySequence_List(pyargs);
+    if (pyargs == NULL) {
+      goto inner_cleanup;
+    }
+
+    unknown_ros_pyargs = PyList_New(0);
+    if (unknown_ros_pyargs == NULL) {
+      goto inner_cleanup;
+    }
+
+    for (int i = 0; i < unparsed_ros_args_count; ++i) {
+      PyObject * ros_pyarg = PySequence_GetItem(pyargs, (Py_ssize_t)i);
+      if (ros_pyarg == NULL) {
+        goto inner_cleanup;
+      }
+      if (PyList_Append(unknown_ros_pyargs, ros_pyarg) != 0) {
+        goto inner_cleanup;
+      }
```

Comment:
You're right, but that's because I'm not using the right API. I should be doing `PyList_GetItem` (which borrows a reference) instead of `PySequence_GetItem` (which creates a new reference).